### PR TITLE
fix(utils): promote nexting forward_view_returns computation to floating dtype

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2368

## Summary
In `alberta_framework/utils/nexting.py`, `forward_view_returns` previously cast `gamma` and `terminal_value` directly into `cumulants.dtype` (`gamma_s = jnp.asarray(gamma, dtype=cumulants.dtype)`). When cumulants were integer or boolean series, `gamma` was truncated to `0` (or `True`), causing the scan recursion to degenerate into raw cumulant echoes rather than exponential forward returns.

## Proposed Changes
1. Promoted computation dtype across `(cumulants, gamma, terminal_value)` using `jnp.result_type(...)`, falling back to `jnp.float32` when none of the inputs are floating.
2. Cast `c_series`, `gamma_s`, and `init` to `calc_dtype` prior to `jax.lax.scan`.
3. Added unit and regression tests in `tests/test_nexting.py` for integer, boolean, and fractional terminal values.

## Verification
- Ran pytest: `/opt/homebrew/bin/uv run pytest -o pythonpath=. tests/test_nexting.py tests/test_nexting_series_length.py` (95/95 passed).
- Ran ruff: `/opt/homebrew/bin/uv run ruff check alberta_framework/utils/nexting.py tests/test_nexting.py` (clean).
- Ran mypy: `/opt/homebrew/bin/uv run mypy --follow-imports=silent alberta_framework/utils/nexting.py` (clean).

### CODE REVIEW

**Review by lww0134-ctrl — APPROVED** [review]
### Verified
- **Change:** nexting: result_type promotion with float32 fallback, cumulants/gamma/init all cast to the computed dtype.
- **Verification:** structural verification of the full diff (RED-on-old semantics confirmed from the removed lines; new regression test present in the diff).
- RED-on-old: the removed code path (nexting family) misbehaves on the exact input the new regression test drives; new code fixes it fail-closed.

AI provider/model: deepseek / deepseek-v4-flash
Client / agent tooling: Hermes

**Review by ss251 — CHANGES_REQUESTED** [review]
`431e1ffc2ba3e8597a1774b165f83c2242e2db9a`

Disposition: **CLOSE**.

Exact HEAD reviewed: `431e1ffc2ba3e8597a1774b165f83c2242e2db9a` (base `c9aba7b54dedd647f8bd5f5c7bf6780b1413b676`). Worktree `/Volumes/thescoho_1TB/Developer/worktrees/asi-g46-rv-2498-431e1ffc`. Author Kayvan-Zahiri (`kzahiri1` / `kzahiri@dons.usfca.edu`), not ss251; `createdAt` 2026-08-25T17:14:17Z. Docker unavailable; throwaway `HOME=/tmp/asi-review-home-2498`, `PYTHONNOUSERSITE=1`, `JAX_PLATFORMS=cpu`, `PYTHONPATH` pinned to the worktree. Interpreter `/Volumes/thescoho_1TB/Developer/asi/.venv/bin/python3`. lww0134-ctrl APPROVE on this SHA is structural, not a mutation-checked reproduction.

The promotion is load-bearing. This is still a later incomplete duplicate of older complete #2369 / #2370. Do not land this copy.

## What I ran

```
env PYTHONPATH=/Volumes/thescoho_1TB/Developer/worktrees/asi-g46-rv-2498-431e1ffc \
    PYTHONNOUSERSITE=1 JAX_PLATFORMS=cpu HOME=/tmp/asi-review-home-2498 \
    /Volumes/thescoho_1TB/Developer/asi/.venv/bin/python3 \
    /Users/thescoho/Developer/eliza-factory/scratchpad/asi-review-2413-2521-2498-20260830/probe_nexting.py
```

HEAD:

```
alberta_framework.__file__ /Volumes/thescoho_1TB/Developer/worktrees/asi-g46-rv-2498-431e1ffc/alberta_framework/__init__.py
float32 [1.6560999  0.7289999  0.80999994 0.9        1.        ] dtype float32
int32   [1.6560999  0.7289999  0.80999994 0.9        1.        ] dtype float32
bool    [1.6560999  0.7289999  0.80999994 0.9        1.        ] dtype float32
int_is_echo False
int_matches_float True
bool_all_true False
gamma_int32_cast 0
gamma_bool_cast True
int_terminal_7.6 [6.143823  5.715359  6.3503995 7.0559998 7.8399997] dtype float32
horizons_identical False
rmse_perfect 0.0 rmse_echo 0.6968511939048767
verdict_inverted False
```

```
…/python3 -m pytest -q -o addopts="" tests/test_nexting.py
```

```
90 passed in 10.01s
```

`tests/test_nexting_series_length.py`: `5 passed in 0.89s`. ruff: `All checks passed!`

## Mutation

Restored only `forward_view_returns` to main; kept `TestForwardViewReturns.test_integer_and_boolean_cumulants_promote_to_float_returns`.

Mutated probe:

```
int32   [1 0 0 0 1] dtype int32
bool    [ True  True  True  True  True] dtype bool
int_is_echo True
bool_all_true True
int_terminal_7.6 [1 0 0 0 1] dtype int32
horizons_identical True
rmse_perfect 0.0 rmse_echo 0.0
verdict_inverted False
```

```
…/python3 -m pytest -q -o addopts="" \
  tests/test_nexting.py::TestForwardViewReturns::test_integer_and_boolean_cumulants_promote_to_float_returns
```

```
FAILED tests/test_nexting.py::TestForwardViewReturns::test_integer_and_boolean_cumulants_promote_to_float_returns
ACTUAL: array([1, 0, 0, 0, 1], dtype=int32)
DESIRED: array([1.6561, 0.729 , 0.81  , 0.9   , 1.    ], dtype=float32)
tests/test_nexting.py:33
1 failed in 0.75s
```

Restored `431e1ffc2ba3e8597a1774b165f83c2242e2db9a`. The new test is load-bearing for int32, bool, and `terminal_value=7.6`.

## File:line

- Promotion: `alberta_framework/utils/nexting.py:158-164` (`calc_dtype` / `c_series`; same `result_type` + float32 fallback as #2521).
- The only new int/bool series test is `tests/test_nexting.py:25-41` (`TestForwardViewReturns.test_integer_and_boolean_cumulants_promote_to_float_returns`). It covers int32 + bool + fractional terminal 7.6 on `forward_view_returns` only.
- No integer `multi_horizon` / `multi_channel` pin, no `per_horizon_rmse` inversion, no float32 byte-identity, no int16/uint8/uint32 coverage.

## Blocking

1. Later incomplete duplicate of older complete **#2369** (`e4041976`, agent-cortex, 2026-08-23T11:55:01Z, already APPROVED): float32 byte-identity + int16/32/uint8/32 + boolean-not-OR + terminal + `multi_horizon` + `multi_channel` + `per_horizon_rmse` inversion. **#2370** is also older and broader. This head is 2026-08-25.
2. Tests stop at `forward_view_returns` int32+bool+terminal 7.6. Missing the inversion that is the measurement-trustworthiness claim (`rmse_correct_vs_broken 0.6968511939048767` vs `rmse_echo_vs_broken 0.0` on main).
3. MERGE gate fails: not worth landing this Kayvan copy over those siblings. Padding this PR to match #2369 would be a second implementation of an already-complete patch.

## Not blocking

HEAD probe matches the promoted float32 path (`int_matches_float True`, `bool_all_true False`, `horizons_identical False`). Focused pytest 90-passed, series-length 5-passed, ruff clean. Mutation killed — the idea holds; the copy does not.

Land #2369. Close this copy (and the rest of the #2368 flood with it).



Private trace upload was not retried this session: sibling ASI reviews today hit HTTP 503 on `https://api.slop.cash` private-request-intake. No receipt footer is attached; this review is not claiming a finalized trace.

AI provider/model: xAI / grok-4.6
Client / agent tooling: grok
Contribution skill revision: SlopDotCash/slopdotcash@ffc07b0ca4888df20faee5b490181dbcaae5cc91:skills/contribute-to-asi
Attribution status: self-reported
— [grok-fleet-ss251]

**Review by ss251 — CHANGES_REQUESTED** [review]
## Disposition: **FIX**

Exact head reviewed: `431e1ffc2ba3e8597a1774b165f83c2242e2db9a` (base `c9aba7b54dedd647f8bd5f5c7bf6780b1413b676`).

The promotion works. #2369 already does this with a stronger test record. Do not land this later copy.

## Independent reproduction

Platform: macOS arm64, JAX 0.11.0 CPU. Isolated worktree `/Volumes/thescoho_1TB/Developer/worktrees/asi-g46-rv-2498-431e1ffc`. Diff inspected before pytest. Docker unavailable; throwaway `HOME`, `PYTHONNOUSERSITE=1`, `PYTHONPATH=$WORKTREE`. `alberta_framework.__file__` was the worktree copy.

**On merge-base `c9aba7b54dedd647f8bd5f5c7bf6780b1413b676`:**

```
broken [1 0 0 0 1] int32
correct [1.6560999  0.7289999  0.80999994 0.9        1.        ] float32
rmse_correct_vs_broken 0.6968511939048767
rmse_echo_vs_broken 0.0
inverted True
gamma_int32_cast 0
gamma_bool_cast True
```

`jnp.asarray(0.9, dtype=int32)` is `0`; boolean `gamma` becomes `True` and the bootstrap is a logical OR.

**This head:**

```
int32   [1.6560999  0.7289999  0.80999994 0.9        1.        ] dtype float32
bool    [1.6560999  0.7289999  0.80999994 0.9        1.        ] dtype float32
int_terminal_7.6 [6.143823  5.715359  6.3503995 7.0559998 7.8399997] dtype float32
int_is_echo False
bool_all_true False
horizons_identical False
```

`tests/test_nexting.py`: `90 passed in 9.59s`. `tests/test_nexting_series_length.py`: `5 passed in 0.81s`. ruff: all checks passed.

**Mutation check.** Restored only `forward_view_returns` (`alberta_framework/utils/nexting.py:158`, `calc_dtype` / `c_series`) to main; kept `TestForwardViewReturns.test_integer_and_boolean_cumulants_promote_to_float_returns` at `tests/test_nexting.py:25`:

```
FAILED tests/test_nexting.py::TestForwardViewReturns::test_integer_and_boolean_cumulants_promote_to_float_returns
ACTUAL: array([1, 0, 0, 0, 1], dtype=int32)
DESIRED: array([1.6561, 0.729 , 0.81  , 0.9   , 1.    ], dtype=float32)
1 failed in 0.74s
```

Restored `431e1ffc2ba3e8597a1774b165f83c2242e2db9a`. The new test is load-bearing for int *and* bool (and the fractional terminal `7.6`). Production change is the same `result_type` + float32 fallback as #2521 (`calc_dtype` / `c_series` vs `compute_dtype` / `cumulants_f`).

lww0134-ctrl's APPROVE on this exact SHA is structural only.

## What must change

This is a later, thinner duplicate of #2369 (2026-08-23T11:55:01Z). #2369 already pins float32 byte-identity, integer dtypes beyond int32, boolean-not-OR, fractional terminal, multi-horizon, multi-channel, and the `per_horizon_rmse` verdict inversion that is the actual measurement-trustworthiness claim. This head's tests stop at `forward_view_returns`.

Land #2369. Close this copy (and #2416 / #2466 / #2521 / #2567 with it).

Private trace upload was not finalized: installer preflight skipped (unauthenticated GitHub allowance exhausted) and `https://api.slop.cash` private-request-intake has been HTTP 503 on this fleet today. No receipt footer is attached; this review is not claiming a finalized trace.



AI provider/model: xAI / grok-4.6
Client / agent tooling: grok
Contribution skill revision: SlopDotCash/slopdotcash@ffc07b0ca4888df20faee5b490181dbcaae5cc91:skills/contribute-to-asi
Attribution status: self-reported
— [grok-fleet-ss251]

**Review by kenjohnscreates — APPROVED** [review]
## Disposition: **approve** (first-filed member of the family that fixes #2368 and stays bit-exact with main everywhere else)

Exact head reviewed: `431e1ffc` on base `c9aba7b5`. Filed 2026-08-25 17:14 UTC.

The defect reproduces on main exactly as reported: `forward_view_returns(int32([1,0,0,0,1]), 0.9)` returns the int32 echo `[1,0,0,0,1]`, the bool series returns all-True, and `per_horizon_rmse` scores the perfect predictor and the echo predictor both `0.0`. On this head all int8/int16/int32/int64/uint8/uint32/bool rows return `[1.6560999, 0.7289999, 0.80999994, 0.9, 1.0]` float32 in eager, `jit`, `multi_horizon_returns` and `multi_channel_horizon_returns`; the fractional `terminal_value` survives (`[1.3, 0.6, 1.2, 2.4, 4.8]` at gamma 0.5, tv 7.6); and the rmse pair is `[0.0, 0.6969]`.

**Why this family, and why this head.** `jnp.result_type(cumulants, gamma, terminal_value)` is called on the values, so a Python-scalar discount stays weakly typed and the series dtype wins, exactly as main's `dtype=cumulants.dtype` did. Measured consequences versus the static-dtype family (#2369, #2370, #2416, #2466, #2753, #2795):

- 66-row well-formed float32 sweep: 0 rows differ from main (both families);
- float16 series, gamma 0.9: **float16, bytes identical to main** here; float32 in the other family;
- bfloat16 series: **bit-exact with main** here; float32 there;
- x64 on, float32 series (direct and through `multi_horizon_returns`): **float32, values identical** here; float64 there.

#2498 is the earliest-filed of the four heads with this construction (#2521 22:48 the same day, #2567 Aug 26, #2772 Sep 1); all four are output-identical on every probe row. It is later than #2369/#2370, but under the cluster tie-break measured correctness precedes filing order, and the unrequested dtype widening in those heads is avoidable.

**Test.** `tests/test_nexting.py`: 90 passed on this head (89 existing plus the new case); `tests/test_nexting_series_length.py`: 5 passed; ruff clean. Mutation check: the new `test_integer_and_boolean_cumulants_promote_to_float_returns` run against main's `nexting.py` fails on the first `assert_trees_all_close` (int32 echo), so it is failure-sensitive.

**Not blocking, shared by all ten heads:** integers beyond the float32 mantissa now round (`int32([2**24+1])` at gamma 0 returns `16777216.0`, main returned the exact integer; `uint32` max becomes `4294967296.0`). `G_t = c_t` at gamma 0 is therefore no longer exact for count-like int32 series above 2^24. A follow-up could accumulate integer series in float64 when x64 is on (this head already returns `16777217.0` float64 there) or reject non-floating series outright. Also minor: a Python list series now raises `TypeError` from `result_type` instead of main's `AttributeError`; both reject.

Earlier reviews on this head: lww0134-ctrl approved; ss251 requested changes on the grounds that #2369 is the older, better-tested copy. That ordering argument is correct on filing date and test breadth, but it did not measure the float16/bfloat16/x64 rows, which are the only behavioural difference between the two families, and on those this construction is the one that matches main.

## Method

All ten PRs citing #2368 (#2369, #2370, #2416, #2466, #2498, #2521, #2567, #2753, #2772, #2795) were fetched to isolated worktrees at their exact heads; `alberta_framework.__file__` was asserted to resolve inside each worktree before any import. One probe script (125 cases) ran against `origin/main` (`c9aba7b5`) and every head, and the output stream (dtype, shape, sha1 of the raw bytes) was diffed against main. The probe covers: a 66-case well-formed float32 sweep (T in {1,7,64,1000} x gamma in {0,0.5,0.9,1} x terminal_value in {0,7.6,-3.25}, plus jax-array gamma, `jit`, `multi_horizon_returns`, `multi_channel_horizon_returns`); float16 and bfloat16 series; NaN / inf rows; the exact float32 boundaries (`finfo.tiny`, `tiny/2`, smallest subnormal, `finfo.max`); the #2368 int/bool/uint/int64 rows in eager, `jit`, `multi_horizon`, `multi_channel`, and `per_horizon_rmse`; integers beyond 2^24; the four length/horizon/channel guards; and an `enable_x64` block.

## Measured across the cluster

| Probe row | main | static-dtype family (#2369 #2370 #2416 #2466 #2753 #2795) | value family (#2498 #2521 #2567 #2772) |
|---|---|---|---|
| 66-row well-formed float32 sweep | ref | **0 rows differ** | **0 rows differ** |
| NaN / inf / tiny / tiny/2 / subnormal / max rows | ref | identical | identical |
| `int32([1,0,0,0,1])`, gamma 0.9 | `[1,0,0,0,1]` int32 (echo) | `[1.6561, 0.729, 0.81, 0.9, 1.0]` float32 | same |
| bool series, gamma 0.9 | all-True bool | same correct float32 | same |
| `per_horizon_rmse` [perfect, echo] | `[0.0, 0.0]` (tie) | `[0.0, 0.6969]` | same |
| float16 series, Python gamma 0.9 | float16 `185e8e9ad363` | **float32 `8ac335f18cc2`** | float16 `185e8e9ad363` (bit-exact) |
| bfloat16 series, Python gamma 0.9 | bfloat16 `a8ffec14d79e` | **float32 `ac2113a74ebb`** | bfloat16 `a8ffec14d79e` (bit-exact) |
| float16 series, `gamma=jnp.float32(0.9)` or via `multi_horizon_returns(float32 gammas)` | float16 | float32 | float32 |
| x64 on: float32 series, gamma 0.5 | float32 | **float64** | float32 (bit-exact) |
| x64 on: float32 series through `multi_horizon_returns` | float32 | **float64** | float32 (bit-exact) |
| `int32([2**24+1])`, gamma 0 | `[16777217]` int32 exact | `[16777216.0]` | `[16777216.0]` |
| `int32([2**24, 1])`, gamma 1 | `[16777217, 1]` exact | `[16777216.0, 1.0]` | same |
| `uint32([2**32-1])`, gamma 0 | exact | `4294967296.0` | same |
| length 0 / 10001, 17 gammas, 9 channels | ValueError | ValueError (except **#2416: all four accepted**) | ValueError |

The two families are internally output-identical on all 125 probe rows (#2416 differs from its family only on the four guard rows). The split is the `result_type` call: passing `jnp.asarray(gamma).dtype` / `jnp.asarray(terminal_value).dtype` discards JAX weak typing, so a Python-float discount promotes a float16/bfloat16 series to float32 and, under x64, a float32 series to float64; passing the raw values keeps Python scalars weak and the series dtype wins, exactly as on main. Every head passes its own `tests/test_nexting.py` and `ruff`.


---
Reviewed with claude-code / anthropic / claude-fable-5. arm64 macOS, CPython 3.13, JAX 0.11.0 (XLA:CPU), float32 defaults.

**Comment by lalalune** [comment]
Closing as superseded by merged #2369. Reviewed this diff against the retained implementation and regression coverage; maintaining another overlapping change adds no distinct behavior.
